README.md
# Water Drop 💦💦💦  

[![Pub](https://img.shields.io/pub/v/water_drop.svg)](https://pub.dartlang.org/packages/water_drop)

A simple package for adding water drops to your Widgets!  

👇👇👇👇👇👇👇👇👇👇  
Blog article with detailed explanation:  
https://fidev.io/water-drop  
👆👆👆👆👆👆👆👆👆👆

![Demo](https://user-images.githubusercontent.com/16286046/77296593-412e7e80-6ce8-11ea-9a79-cb3909070886.gif) 

## Why?

Why not?!

## How?

Just wrap your widget in `WaterDrop` and provide `params`:
```dart
Widget build(BuildContext context) {
  return WaterDrop(
    child: _StaticCard(color: Colors.green),
    params: [
      WaterDropParam(top: 50, height: 70, left: 100, width: 50),
      WaterDropParam(top: 10, height: 100, left: 280, width: 100),
      WaterDropParam(top: 155, height: 35, left: 135, width: 35),
      WaterDropParam(top: 135, height: 40, left: 250, width: 30),
      WaterDropParam(top: 20, height: 40, left: 20, width: 40),
      WaterDropParam(top: 140, height: 50, left: 15, width: 40),
      WaterDropParam(top: 20, height: 30, left: 200, width: 30),
      WaterDropParam(top: 120, height: 20, left: 360, width: 20),
    ],
  );
}
```

Or if you want a single drop, use `WaterDrop.single`:
```dart
@override
Widget build(BuildContext context) {
  return WaterDrop.single(
    top: 50,
    height: 70,
    left: 100,
    width: 50,
    child: _StaticCard(color: Colors.green),
  );
}
```


CHANGELOG.md
## [1.0.0] - 23.03.2020

* Initial release

lib/water_drop.dart
library water_drop;

import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

///Parameter of a single water drop
class WaterDropParam {
  ///Distance from the top of child
  final double top;

  ///Distance from the left of child
  final double left;

  ///Width of a water drop
  final double width;

  ///Height of a water drop
  final double height;

  WaterDropParam(
      {@required this.top,
      @required this.left,
      @required this.width,
      @required this.height})
      : assert(top != null),
        assert(left != null),
        assert(width != null && width > 0),
        assert(height != null && height > 0);
}

class WaterDrop extends StatelessWidget {
  ///Parameters of water drops. Each parameter represents one drop.
  final List<WaterDropParam> params;

  ///Child on which drops will be drawn.
  final Widget child;

  const WaterDrop({
    Key key,
    @required this.params,
    @required this.child,
  })  : assert(params != null),
        assert(child != null),
        super(key: key);

  ///A factory for creating a single drop
  factory WaterDrop.single(
          {Key key,
          double left,
          double top,
          double height,
          double width,
          Widget child}) =>
      WaterDrop(
        child: child,
        params: [
          WaterDropParam(top: top, left: left, width: width, height: height),
        ],
      );

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: <Widget>[
        child,
        ...params.map((param) => _WaterDrop(
              child: child,
              left: param.left,
              top: param.top,
              width: param.width,
              height: param.height,
            )),
      ],
    );
  }
}

class _WaterDrop extends StatefulWidget {
  final double top;
  final double left;
  final double width;
  final double height;
  final Widget child;

  const _WaterDrop(
      {Key key, this.child, this.top, this.width, this.left, this.height})
      : super(key: key);

  @override
  __WaterDropState createState() => __WaterDropState();
}

class __WaterDropState extends State<_WaterDrop> {
  ///Size of a child widget. It is needed to provide correct gradient on the drop.
  Size totalSize;

  @override
  Widget build(BuildContext context) {
    //After build, rebuild it again but save the size of a widget (see [totalSize])
    if (totalSize == null) {
      SchedulerBinding.instance.addPostFrameCallback((_) {
        setState(() => totalSize = context.size);
      });
    }
    //If there's no totalSize, use size of the screen
    final size = totalSize ?? MediaQuery.of(context).size;

    //Alignment of center
    final alignment = getAlignment(size);

    //used for determining alignments for gradient
    final alignmentModifier = Alignment(
      widget.width / size.width,
      widget.height / size.height,
    );

    //A child with gradient on it for light illusion
    Widget childWithGradient = Container(
      foregroundDecoration: BoxDecoration(
        gradient: LinearGradient(
          begin: alignment - alignmentModifier,
          end: alignment + alignmentModifier,
          colors: [Colors.black, Colors.white],
        ),
        backgroundBlendMode: BlendMode.overlay,
      ),
      child: widget.child,
    );

    return IgnorePointer(
      child: Stack(
        children: <Widget>[
          _OvalShadow(
            width: widget.width,
            height: widget.height,
            top: widget.top,
            left: widget.left,
          ),
          ...List.generate(8, (i) {
            return Transform.scale(
              scale: 1 + 0.02 * i,
              alignment: alignment,
              child: ClipPath(
                clipper: OvalClipper(
                  center: center,
                  width: widget.width * (1 - 0.04 * i),
                  height: widget.height * (1 - 0.04 * i),
                ),
                clipBehavior: Clip.hardEdge,
                child: childWithGradient,
              ),
            );
          }),
          _LightDot(
            width: widget.width,
            height: widget.height,
            top: widget.top,
            left: widget.left,
          ),
        ],
      ),
    );
  }

  ///Get center of a drop
  Offset get center => Offset(
        widget.left + widget.width / 2,
        widget.top + widget.height / 2,
      );

  ///Map Center and Size to Alignment
  Alignment getAlignment(Size size) => Alignment(
        (center.dx - size.width / 2) / (size.width / 2),
        (center.dy - size.height / 2) / (size.height / 2),
      );
}

class OvalClipper extends CustomClipper<Path> {
  final double height;
  final double width;
  final Offset center;

  OvalClipper({this.center, this.width, this.height});

  @override
  Path getClip(Size size) {
    Path path = Path()
      ..addOval(Rect.fromCenter(
        center: center,
        width: width,
        height: height,
      ));
    return path;
  }

  @override
  bool shouldReclip(CustomClipper<Path> oldClipper) => true;
}

///A white dot in top left corner
class _LightDot extends StatelessWidget {
  final double top;
  final double left;
  final double width;
  final double height;

  const _LightDot({Key key, this.top, this.left, this.width, this.height})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Positioned(
      left: left + width / 4,
      top: top + height / 4,
      width: width / 4,
      height: height / 4,
      child: DecoratedBox(
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          boxShadow: [
            BoxShadow(
              blurRadius: 3,
              color: Colors.white.withOpacity(0.9),
            ),
          ],
        ),
      ),
    );
  }
}

///A shadow below the drop
class _OvalShadow extends StatelessWidget {
  final double top;
  final double left;
  final double width;
  final double height;

  const _OvalShadow({Key key, this.top, this.left, this.width, this.height})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Positioned(
      left: left,
      top: top,
      width: width,
      height: height,
      child: DecoratedBox(
        decoration: BoxDecoration(
          borderRadius: BorderRadius.all(
            Radius.elliptical(width / 2, height / 2),
          ),
          boxShadow: [
            BoxShadow(
              blurRadius: 4,
              offset: Offset(4, 4),
              color: Colors.black.withOpacity(0.5),
            ),
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:water_drop/water_drop.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Water drop',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: WaterDrop(
          params: [
            WaterDropParam(top: 6, height: 40, left: 6, width: 60),
            WaterDropParam(top: 16, height: 30, left: 100, width: 20),
          ],
          child: Container(
            color: Colors.blue,
            padding: const EdgeInsets.all(16),
            child: Text('Water drop'),
          ),
        ),
      ),
      body: SingleChildScrollView(
        child: Column(
          children: <Widget>[
            MyCard(color: Colors.green),
            MyCard(color: Colors.yellow),
            MyCard(color: Theme.of(context).cardColor),
            MyCard(color: Colors.orange),
          ],
        ),
      ),
    );
  }
}

class MyCard extends StatefulWidget {
  final Color color;

  const MyCard({Key key, this.color = Colors.green}) : super(key: key);

  @override
  _MyCardState createState() => _MyCardState();
}

class _MyCardState extends State<MyCard> with SingleTickerProviderStateMixin {
  AnimationController _animationController;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 1000),
    )..repeat(reverse: true);
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animationController,
      builder: (context, child) {
        final move = _animationController.value * 40;
        return WaterDrop(
          child: child,
          params: [
            WaterDropParam(top: 50 + move, height: 70, left: 100, width: 50),
            WaterDropParam(top: 10 + move, height: 100, left: 280, width: 100),
            WaterDropParam(top: 155 + move, height: 35, left: 135, width: 35),
            WaterDropParam(top: 135 + move, height: 40, left: 250, width: 30),
            WaterDropParam(top: 20 + move, height: 40, left: 20, width: 40),
            WaterDropParam(top: 140 + move, height: 50, left: 15, width: 40),
            WaterDropParam(top: 20 + move, height: 30, left: 200, width: 30),
            WaterDropParam(top: 120 + move, height: 20, left: 360, width: 20),
          ],
        );
      },
      child: _StaticCard(color: widget.color),
    );
  }
}

class _StaticCard extends StatelessWidget {
  final Color color;

  const _StaticCard({Key key, this.color = Colors.green}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Card(
      color: color,
      child: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          children: <Widget>[
            Text(
                'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.'),
            RaisedButton(
              color: Colors.blue,
              textColor: Colors.white,
              child: Text('Press me'),
              onPressed: () {},
            )
          ],
        ),
      ),
    );
  }
}
